flink-tests/src/test/java/org/apache/flink/test/checkpointing/SavepointFormatITCase.java

```diff
@@ -78,138 +82,151 @@
+            return stateBackendConfig;
+        }
 
         public int getCheckpointsBeforeSavepoint() {
             return 0;
         }
 
+        protected abstract String getConfigName();
+
         @Override
         public final String toString() {
-            return getName();
+            return String.format(
+                    "%s, incremental: %b, changelog: %b", getName(), incremental, changelogEnabled);
         }
-    }
 
-    private static final StateBackendConfig HEAP =
-            new StateBackendConfig() {
-                @Override
-                public String getName() {
-                    return "HEAP";
-                }
-
-                @Override
-                public Configuration getConfiguration() {
-                    Configuration stateBackendConfig = new Configuration();
-                    stateBackendConfig.setString(StateBackendOptions.STATE_BACKEND, "filesystem");
-                    stateBackendConfig.set(
-                            CheckpointingOptions.FS_SMALL_FILE_THRESHOLD, MemorySize.ZERO);
-                    return stateBackendConfig;
-                }
-            };
-
-    private static final StateBackendConfig ROCKSDB_FULL_SNAPSHOTS =
-            new StateBackendConfig() {
-                @Override
-                public String getName() {
-                    return "ROCKSDB_FULL_SNAPSHOTS";
-                }
+        private static final List<BiFunction<Boolean, Boolean, StateBackendConfig>> builders =
+                asList(SavepointFormatITCase::getRocksdb, SavepointFormatITCase::heap);
 
-                @Override
-                public Configuration getConfiguration() {
-                    Configuration stateBackendConfig = new Configuration();
-                    stateBackendConfig.setString(StateBackendOptions.STATE_BACKEND, "rocksdb");
-                    stateBackendConfig.set(
-                            CheckpointingOptions.FS_SMALL_FILE_THRESHOLD, MemorySize.ZERO);
-                    stateBackendConfig.set(CheckpointingOptions.INCREMENTAL_CHECKPOINTS, false);
-                    return stateBackendConfig;
-                }
-            };
+        public abstract boolean isIncremental();
 
-    private static final StateBackendConfig ROCKSDB_INCREMENTAL_SNAPSHOTS =
-            new StateBackendConfig() {
-                @Override
-                public String getName() {
-                    return "ROCKSDB_INCREMENTAL_SNAPSHOTS";
-                }
+        private boolean isChangelogEnabled() {
+            return changelogEnabled;
+        }
+    }
 
-                @Override
-                public int getCheckpointsBeforeSavepoint() {
-                    return 1;
-                }
+    private static StateBackendConfig heap(boolean incremental, boolean changelogEnabled) {
+        return new StateBackendConfig(changelogEnabled, incremental /* ignored for now */) {
```

Comment:
Does it mean we run the same tests for `HEAP` backend twice? (twice for native, and twice for canonical) Shall we skip one execution?

Reply:
Yes, heap incremental is the same as heap non-incremental.

I could add `assumeFalse(incremental)`; but then adding incremental checkpoints would be more difficult.
The proper way probably is to have `supportsIncrementalCheckpoints` in `StateBackend`; but it 1) complicates the interface; b) requires this test to instantiate backend instance instead of using a label.

Locally, skipping these tests saves several seconds. So I'd rather keep it as is.
WDYT?

Reply:
I'll leave it up to you.

Reply:
I'll leave it as is then.